File: app/src/main/java/com/veyndan/paper/reddit/post/media/mutator/LinkImageMutatorFactory.java
package com.veyndan.paper.reddit.post.media.mutator;

import com.veyndan.paper.reddit.api.reddit.model.Source;
import com.veyndan.paper.reddit.post.media.model.LinkImage;
import com.veyndan.paper.reddit.post.model.Post;

import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.Single;

final class LinkImageMutatorFactory implements MutatorFactory {

    @Override
    public Maybe<Post> mutate(final Post post) {
        return Single.just(post)
                .filter(post1 -> post1.preview().images.size() > 0)
                .map(post1 -> {
                    final Source source = post1.preview().images.get(0).source;
                    final LinkImage linkImage = LinkImage.create(source.url, post1.domain());
                    return post1.withMedias(post1.medias().value.concatWith(Observable.just(linkImage)));
                });
    }
}

File: app/src/main/java/com/veyndan/paper/reddit/post/media/mutator/LinkImageMutatorFactory.kt
package com.veyndan.paper.reddit.post.media.mutator

import com.veyndan.paper.reddit.api.reddit.model.Source
import com.veyndan.paper.reddit.post.media.model.LinkImage
import com.veyndan.paper.reddit.post.model.Post
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Single

class LinkImageMutatorFactory : MutatorFactory {

    override fun mutate(post: Post): Maybe<Post> {
        return Single.just(post)
                .filter {it.preview.images.isNotEmpty() }
                .map {
                    val source: Source = it.preview.images[0].source
                    val linkImage: LinkImage = LinkImage(source.url, it.domain!!)
                    it.copy(it.medias.concatWith(Observable.just(linkImage)))
                }
    }
}
